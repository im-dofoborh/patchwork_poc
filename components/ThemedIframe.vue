<template>
  <iframe
    ref="iframe"
    :srcdoc="srcdoc"
    :style="style"
    :class="classNames"
    scrolling="no"
    @load="onLoad" />
</template>

<script>
  export default {
    props: {
      theme: {
        type: Object,
        required: true,
      },
      markup: {
        type: String,
        default: '',
      },
      wrapInDocument: {
        type: Boolean,
        default: true
      },
      bodyClassNames: {
        type: String,
        default: '',
      },
      size: {
        type: String,
        default: '',
      }
    },

    data() {
      return {
        style: {
          height: 0,
        }
      }
    },

    computed: {
      srcdoc() {
        const { css, vendorJs, appJs, grunticon, webpackChunkManifest, dist, typekitId } = this.theme;

        if (!this.wrapInDocument) {
          return this.markup;
        }

        return `
          <!doctype>
          <html>
            <head>
              <link rel="stylesheet" type="text/css" href="${css}">
              <script>
                window.webpackManifest = ${webpackChunkManifest};
                window.chunkURL = '${dist}/'; // For dynamic cross-domain importing of chunks
              <\/script>
            </head>
            <body class="${this.bodyClassNames}" style="background-color: transparent;">
              ${this.markup}
              <script type="text/javascript" src="${vendorJs}"><\/script>
              <script type="text/javascript" src="${appJs}"><\/script>
              <script type="text/javascript" src="${grunticon.loader}"><\/script>
              <script>
                window.onload = function () {
                  grunticon(
                    ${JSON.stringify(grunticon.stylesheets)},
                    grunticon.svgLoadedCORSCallback
                  );
                };
              <\/script>
              <script>
                (function(d) {
                  var config = {
                    kitId: '${typekitId}',
                    scriptTimeout: 3000,
                    async: true
                  },
                  h=d.documentElement,t=setTimeout(function(){h.className=h.className.replace(/\bwf-loading\b/g,"")+" wf-inactive";},config.scriptTimeout),tk=d.createElement("script"),f=false,s=d.getElementsByTagName("script")[0],a;h.className+=" wf-loading";tk.src='https://use.typekit.net/'+config.kitId+'.js';tk.async=true;tk.onload=tk.onreadystatechange=function(){a=this.readyState;if(f||a&&a!="complete"&&a!="loaded")return;f=true;clearTimeout(t);try{Typekit.load(config)}catch(e){}};s.parentNode.insertBefore(tk,s)
                })(document);
              <\/script>
            </body>
          </html>
        `;
      },

      classNames() {
        let classNames = 'sg-themed-iframe';
        if (this.size) {
          classNames += ` sg-themed-iframe--${this.size.toLowerCase()}`;
        }
        return classNames;
      }
    },

    methods: {
      /**
       * Handle the iframe load event.
       */
      onLoad() {
        if (!('MutationObserver' in window)) {
          this.setHeightAtIntervals(400);
          return;
        }
        this.setHeight();
        this.setHeightOnMutation();
        this.$emit('load');
      },

      /**
       * Set the iframe height at given intervals.
       * @param {number} interval
       *   An interval in milliseconds.
       * @param {number} clearAfter
       *   Milliseconds after which to stop calling.
       */
      setHeightAtIntervals(interval, clearAfter) {
        setInterval(this.setHeight, interval);
        if (clearAfter) {
          setInterval(this.setHeight, interval);
          setTimeout(() => {
            clearInterval(this.setHeight);
          }, clearAfter);
        }
      },

      /**
       * Set iframe height when mutation observed.
       */
      setHeightOnMutation() {
        const { iframe } = this.$refs;
        if (!iframe) {
          return;
        }
        const config = {
          attributes: true,
          childList: true,
          subtree: true,
        };
        const observer = new MutationObserver(() => {
          // Update height for 5 seconds to allow for any animations
          this.setHeightAtIntervals(24, 5000);
        });
        observer.observe(iframe.contentWindow.document.body, config);
      },

      /**
       * Set the height of the iframe.
       */
      setHeight() {
        const { iframe } = this.$refs;
        if (!iframe) {
          return;
        }
        const height = this.getHeightOfTallestChild(iframe.contentWindow.document.body);
        this.style.height = `${height}px`;
      },

      /**
       * Get the height of the tallest child inside a root element.
       * @param {Element} root
       *   The root element.
       * @returns {number}
       *  The pixel height of the tallest child.
       */
      getHeightOfTallestChild(root) {
        let maxHeight = -1;
        this.forAllVisibleChildNodes(root, (element, computedStyle) => {
          const boundingRect = element.getBoundingClientRect();
          maxHeight = Math.max(maxHeight, boundingRect.top
            + boundingRect.height
            + Math.abs(parseFloat(computedStyle.marginBottom))
            + this.getShadowHeight(computedStyle)
          );
        });
        return maxHeight;
      },

      /**
       * Get the height of any box shadows.
       * @param {object} computedStyle
       *   The CSS properties of an element.
       * @returns {number}
       *   The pixel height of any shadows.
       */
      getShadowHeight(computedStyle) {
        if (computedStyle.boxShadow !== 'none') {
          const shadowChunks = computedStyle.boxShadow.split('px ');
          return Math.abs(parseFloat(shadowChunks[1])) // height
            + Math.max(0, parseFloat(shadowChunks[2])) // shadow blur
            + Math.max(0, parseFloat(shadowChunks[3])); // shadow spread
        }
        return 0;
      },

      /**
       * Exacute a callback function for all visible child nodes.
       * @param {Element} node
       *   A root node.
       * @param {Function} cb
       *   A callback function.
       */
      forAllVisibleChildNodes(node, cb) {
        if (node && node.nodeType === 1) {
          const computedStyle = window.getComputedStyle(node);
          cb(node, computedStyle);
          if (computedStyle.getPropertyValue('overflow') === 'visible') {
            for (const child of node.childNodes) {
              this.forAllVisibleChildNodes(child, cb);
            }
          }
        }
      }
    },
  }
</script>

<style lang="scss" scoped>
  .sg-themed-iframe {
    display: block;
    border: 0;
    height: 0;
    transition: width .3s ease;
    width: 100%;
    max-width: 100%;

    &--xs {
      width: map-get($grid-breakpoints, sm) - 1;
    }

    &--sm {
      width: map-get($grid-breakpoints, md) - 1;
    }

    &--md {
      width: map-get($grid-breakpoints, lg) - 1;
    }
  }
</style>