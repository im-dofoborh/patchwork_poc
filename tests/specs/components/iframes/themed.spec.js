import { mount, createLocalVue } from '@vue/test-utils';
import ThemedIframe from '~/components/ThemedIframe';
import { theme } from '~/tests/fixtures';

describe('ContentEditor component', () => {
  let localVue = null;
  let wrapper = null;

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.component('themed-iframe', ThemedIframe);
    wrapper = mount(ThemedIframe, {
      localVue,
      propsData: {
        theme,
      }
    });
  })

  it('renders default variant', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders with markup', () => {
    wrapper.setProps({
      markup: '<div class="foo">My Markup</div>',
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders small variant', () => {
    wrapper.setProps({
      size: 'sm',
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders unwrapped variant', () => {
    wrapper.setProps({
      wrapInDocument: false,
      markup: '<div class="foo">My Markup</div>',
    });
    expect(wrapper.html()).toMatchSnapshot();
  });
});
