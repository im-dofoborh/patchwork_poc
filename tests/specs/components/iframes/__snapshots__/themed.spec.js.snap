// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentEditor component renders default variant 1`] = `
<iframe srcdoc="
        <!doctype>
        <html>
          <head>
            <link rel=&quot;stylesheet&quot; type=&quot;text/css&quot; href=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/css/app.css&quot;>
            <script>
              window.webpackManifest = {&quot;0&quot;: &quot;js/0.js&quot;, &quot;1&quot;: &quot;js/1.js&quot;};
              window.chunkURL = 'https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/'; // For dynamic cross-domain importing of chunks
            </script>
          </head>
          <body class=&quot;&quot; style=&quot;background-color: transparent;&quot;>
            
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/js/vendor.js&quot;></script>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/js/app.js&quot;></script>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/grunticon-1f6ae85a3d.loader.js&quot;></script>
            <script>
              window.onload = function () {
                grunticon(
                  [&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-adb1c72d37.data.svg.css&quot;,&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-27488828d3.data.png.css&quot;,&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-61b5c20abf.fallback.css&quot;],
                  grunticon.svgLoadedCORSCallback
                );
              };
            </script>
            <script>
              (function(d) {
                var config = {
                  kitId: 'sts5pxq',
                  scriptTimeout: 3000,
                  async: true
                },
                h=d.documentElement,t=setTimeout(function(){h.className=h.className.replace(/wf-loading/g,&quot;&quot;)+&quot; wf-inactive&quot;;},config.scriptTimeout),tk=d.createElement(&quot;script&quot;),f=false,s=d.getElementsByTagName(&quot;script&quot;)[0],a;h.className+=&quot; wf-loading&quot;;tk.src='https://use.typekit.net/'+config.kitId+'.js';tk.async=true;tk.onload=tk.onreadystatechange=function(){a=this.readyState;if(f||a&amp;&amp;a!=&quot;complete&quot;&amp;&amp;a!=&quot;loaded&quot;)return;f=true;clearTimeout(t);try{Typekit.load(config)}catch(e){}};s.parentNode.insertBefore(tk,s)
              })(document);
            </script>
          </body>
        </html>
      " scrolling="no" class="sg-themed-iframe" style="height: 0px;"></iframe>
`;

exports[`ContentEditor component renders small variant 1`] = `
<iframe srcdoc="
        <!doctype>
        <html>
          <head>
            <link rel=&quot;stylesheet&quot; type=&quot;text/css&quot; href=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/css/app.css&quot;>
            <script>
              window.webpackManifest = {&quot;0&quot;: &quot;js/0.js&quot;, &quot;1&quot;: &quot;js/1.js&quot;};
              window.chunkURL = 'https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/'; // For dynamic cross-domain importing of chunks
            </script>
          </head>
          <body class=&quot;&quot; style=&quot;background-color: transparent;&quot;>
            
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/js/vendor.js&quot;></script>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/js/app.js&quot;></script>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/grunticon-1f6ae85a3d.loader.js&quot;></script>
            <script>
              window.onload = function () {
                grunticon(
                  [&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-adb1c72d37.data.svg.css&quot;,&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-27488828d3.data.png.css&quot;,&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-61b5c20abf.fallback.css&quot;],
                  grunticon.svgLoadedCORSCallback
                );
              };
            </script>
            <script>
              (function(d) {
                var config = {
                  kitId: 'sts5pxq',
                  scriptTimeout: 3000,
                  async: true
                },
                h=d.documentElement,t=setTimeout(function(){h.className=h.className.replace(/wf-loading/g,&quot;&quot;)+&quot; wf-inactive&quot;;},config.scriptTimeout),tk=d.createElement(&quot;script&quot;),f=false,s=d.getElementsByTagName(&quot;script&quot;)[0],a;h.className+=&quot; wf-loading&quot;;tk.src='https://use.typekit.net/'+config.kitId+'.js';tk.async=true;tk.onload=tk.onreadystatechange=function(){a=this.readyState;if(f||a&amp;&amp;a!=&quot;complete&quot;&amp;&amp;a!=&quot;loaded&quot;)return;f=true;clearTimeout(t);try{Typekit.load(config)}catch(e){}};s.parentNode.insertBefore(tk,s)
              })(document);
            </script>
          </body>
        </html>
      " scrolling="no" class="sg-themed-iframe sg-themed-iframe--sm" style="height: 0px;"></iframe>
`;

exports[`ContentEditor component renders unwrapped variant 1`] = `<iframe srcdoc="<div class=&quot;foo&quot;>My Markup</div>" scrolling="no" class="sg-themed-iframe" style="height: 0px;"></iframe>`;

exports[`ContentEditor component renders with markup 1`] = `
<iframe srcdoc="
        <!doctype>
        <html>
          <head>
            <link rel=&quot;stylesheet&quot; type=&quot;text/css&quot; href=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/css/app.css&quot;>
            <script>
              window.webpackManifest = {&quot;0&quot;: &quot;js/0.js&quot;, &quot;1&quot;: &quot;js/1.js&quot;};
              window.chunkURL = 'https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/'; // For dynamic cross-domain importing of chunks
            </script>
          </head>
          <body class=&quot;&quot; style=&quot;background-color: transparent;&quot;>
            <div class=&quot;foo&quot;>My Markup</div>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/js/vendor.js&quot;></script>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-slate-theme/dist/js/app.js&quot;></script>
            <script type=&quot;text/javascript&quot; src=&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/grunticon-1f6ae85a3d.loader.js&quot;></script>
            <script>
              window.onload = function () {
                grunticon(
                  [&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-adb1c72d37.data.svg.css&quot;,&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-27488828d3.data.png.css&quot;,&quot;https://slate.local.wcp.imdserve.com/wcp/themes/im-fabric/dist/icons/icons-61b5c20abf.fallback.css&quot;],
                  grunticon.svgLoadedCORSCallback
                );
              };
            </script>
            <script>
              (function(d) {
                var config = {
                  kitId: 'sts5pxq',
                  scriptTimeout: 3000,
                  async: true
                },
                h=d.documentElement,t=setTimeout(function(){h.className=h.className.replace(/wf-loading/g,&quot;&quot;)+&quot; wf-inactive&quot;;},config.scriptTimeout),tk=d.createElement(&quot;script&quot;),f=false,s=d.getElementsByTagName(&quot;script&quot;)[0],a;h.className+=&quot; wf-loading&quot;;tk.src='https://use.typekit.net/'+config.kitId+'.js';tk.async=true;tk.onload=tk.onreadystatechange=function(){a=this.readyState;if(f||a&amp;&amp;a!=&quot;complete&quot;&amp;&amp;a!=&quot;loaded&quot;)return;f=true;clearTimeout(t);try{Typekit.load(config)}catch(e){}};s.parentNode.insertBefore(tk,s)
              })(document);
            </script>
          </body>
        </html>
      " scrolling="no" class="sg-themed-iframe" style="height: 0px;"></iframe>
`;
